### PR TITLE
Feature/wiki one page

### PR DESCRIPTION
**Pull request by brianjgeiger**
Purpose
-----------
Consolidate the wiki functionality into a single page including new rendering engine and comparison functions. 

Changes
------------
 * Update the rendering engine to allow us to use a single rendering engine everywhere instead of one for preview and one for render. 
 * Rendering engine will allow for syntax extensions. 
 * Comparison done by pluggable diffing engine
 * Comparisons can be done between any two versions, including comparing to live-edited wiki (with write permissions)
 * We introduce new column-based interface to the wiki interface to allow flexibility in what you want to view
 * You can go to a specific URL that matches the columns you want active (Note that URLs may change before we go to production
 * URL updates with changes to column setup, though history is not modified for column changes (on purpose)

Side effects
----------------
People will be very excited by the new interface and want it everywhere?

Known issues
-------------------
 * Mathjax rendering seems to have three states (unrendered, rendered, and rendered but looks different). This is on Production as well, btw. 
 * Syntax extensions (TOC and YouTube) have some matching bugs that need fixing.
 * TOC extension may need some more work (or we're pulling the wrong version, TBD).

This replaces #1833 so Robert doesn't have to keep managing it.

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -10,7 +10,6 @@
     "ace-builds": "~1.1.8",
     "qunit": "~1.14.0",
     "sinon": "~1.10.2",
-    "select2": "~3.5.1",
```
Comment (sloria) [inline bower.json]: Not sure why bower keeps on removing this. Can you add it back?

File: webpack.common.config.js [hunk 2]
```diff
@@ -110,6 +118,9 @@ var resolve = {
         'koHelpers': staticPath('js/koHelpers.js'),
         'addonPermissions': staticPath('js/addonPermissions.js'),
         'navbar-control': staticPath('js/navbarControl.js'),
+        'osf-panel': staticPath('vendor/bower_components/osf-panel/dist/jquery-osfPanel.js'),
```
Comment (sloria) [inline webpack.common.config.js]: This should not be necessary. Since `osf-panel` is bower-installed and has a `main` entry, `require('osf-panel')` will do the right thing.

File: website/static/js/wikiDiff.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+var wikiDiff =  new jsdiff.Diff();
+wikiDiff.tokenize = function (value) {
+    return removeEmpty(value.split(/(\S.+?[.!?\n])(?=\s+|$)/));
+};
+
+var diff = function(beforeText, afterText) {
+    var diffList = wikiDiff.diff(beforeText, afterText);
+    var fragment = document.createDocumentFragment();
+
+    diffList.forEach(function(part) {
```
Comment (sloria) [inline website/static/js/wikiDiff.js]: Since performance is critical here, we might want to change this to a plain `for` loop, which performs faster (http://jsperf.com/foreach-vs-loop).

File: bower.json [hunk 4]
```diff
@@ -26,12 +25,19 @@
...
     "jquery.cookie": "~1.4.1",
     "hgrid": "~0.2.10",
     "uri.js": "~1.14.1",
     "xhook": "~1.3.0",
+    "osf-panel": "https://github.com/caneruguz/osf-panel.git#develop",
+    "markdown-it": "~3.0.5",
+    "markdown-it-sanitizer": "~0.2.0",
+    "markdown-it-video": "https://github.com/brianjgeiger/markdown-it-video.git#b8b1cccb505a6773c30f8919a31b34092bac8e60",
+    "markdown-it-toc": "https://github.com/samchrisinger/markdown-it-toc.git#8c7fb8bfd25ba8b4fadd12aceb923f3066899743",
+    "jsdiff": "~1.2.2",
```
Comment (sloria) [inline bower.json]: Prefer npm distributions to bower components. There is one available for `jsdiff`. You should just have to move this to `package.json`.

File: website/static/js/wikiDiff.js (hunk 3 above)
Comment (brianjgeiger) [inline website/static/js/wikiDiff.js]: I don't know that it will be in this case. The comment at the bottom suggests, "If you can keep the contents of the for loop simple, then consider it. but if the contents get complex such that it would be better implemented in a function, then the forEach loop probably runs about equal and is certainly more readable." And the speed for this is going to be in touching the dom rather than iterating through the loop. With more items and less interior code, I am sure it would be, but in this case, I have my doubts.

**Comment by sloria** [comment]
The diffing between the preview and "Current" version seems to be incorrect.

### Steps

- Create a new wiki page
- Enter some text
- Save
- Enable the "Compare" column
- In the "View" column, switch the current version to "Preview"

### Expected

The "Compare" column should show no changes

### Actual

The "Comparison" column shows that every line was removed and added on the previous line.

osf___wiki_redesign_wiki

**Comment by sloria** [comment]
Is there any way we can have youtube videos scale properly within columns? With more than 2 columns, the right side of the videos get cut off.

osf___wiki_redesign_wiki

**Comment by sloria** [comment]
We may want to add snippets for the new markdown extensions (youtube, toc). You can add the snippets here: https://github.com/brianjgeiger/osf.io/blob/d1d05aa38ab877df1973ae29533dc00e4b51505f/website/addons/wiki/static/ace-markdown-snippets.js

File: website/addons/wiki/static/wikiPage.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    };
+
+    if (typeof self.editor !== 'undefined') {
+        self.editor.on('change', function () {
+            // Quick render
+            self.allowFullRender(false);
+            // Full render
+            clearTimeout(self.renderTimeout);
+
+            self.renderTimeout = setTimeout(function () {
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: You should be able to use `$osf.throttle` here. Something like:

```js
var throttledAllowFullRender = $osf.throttle(self.allowFullRender.bind(self, true), THROTTLE);
```

File: website/addons/wiki/static/wikiPage.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            if (self.version() === 'preview') {
+                self.rendered(self.renderMarkdown(self.viewText()));
+                self.displaySource(self.viewText());
+            } else {
+                if (self.version() === 'current') {
+                    requestURL = contentURL;
+                } else {
+                    requestURL= contentURL + self.version();
+                }
+                var request = $.ajax({
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: Does this work need to happen if `self.visible() === false`?

Comment (brianjgeiger) [inline website/addons/wiki/static/wikiPage.js]: If you're just asking about the ajax call, then it won't happen if it's not visible because there'll be no way to change the version. It would still need to load the initial version if there were a way to, say, instantiate the page with a compare column but not a view column (which I don't think is currently possible, but might be before the end of the week, depending). So I think the ajax request should happen regardless. 

If we want rendering not to happen when it's not visible (which is sane), we'd probably want to add the visibility check to the rendering function.

File: website/addons/wiki/static/wikiPage.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    var self = this;
+    self.version = version;
+    self.viewText = viewText; // comes from EditWidget.viewText
+    self.rendered = rendered;
+    self.visible = visible;
+    self.allowMathjaxification = allowMathjaxification;
+    self.editor = editor;
+    self.allowFullRender = allowFullRender;
+    self.renderTimeout = null;
+    self.displaySource = ko.observable('');
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: What is the purpose for having both `displaySource` and `displayText`?

File: website/addons/wiki/static/wikiPage.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
+'use strict';
+var ko = require('knockout');
+var $ = require('jquery');
+var $osf = require('osfHelpers');
+
+var mathrender = require('mathrender');
+var md = require('markdown').full;
+var mdQuick = require('markdown').quick;
+var diffTool = require('wikiDiff');
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: Perhaps we should rename `wikiDiff.js` to `diffTool.js` to make it more generic.

File: website/addons/wiki/static/wikiPage.js (hunk 7 above)
Comment (brianjgeiger) [inline website/addons/wiki/static/wikiPage.js]: displayText is a computedObservable but needs to do ajax. To get around the limitations of that, I use displaySource as the observable that will be used in later computedObservables.

**Comment by brianjgeiger** [comment]
Pinging @rliebz about https://github.com/CenterForOpenScience/osf.io/pull/1858#issuecomment-75583500

**Comment by brianjgeiger** [comment]
For https://github.com/CenterForOpenScience/osf.io/pull/1858#issuecomment-75585375 … mmmmaybe. Yes, it bothers me a bit as well. I don't think it's the worst thing ever, because the times when you would have all the columns open, you'd probably not be using it for watching the videos. But it can happen at various screen sizes as well, so I'll poke around at the Youtube docs and see if there's something that can be done (or maybe see if @caneruguz has a moment to check).

File: website/addons/wiki/static/wikiPage.js (hunk 5 above)
Comment (brianjgeiger) [inline website/addons/wiki/static/wikiPage.js]: I am unclear as to the benefits of the helper function.

**Comment by brianjgeiger** [comment]
[This](http://css-tricks.com/NetMag/FluidWidthVideo/Article-FluidWidthVideo.php) looks promising for the video width. I should be able to integrate something like that.

File: website/addons/wiki/static/wikiPage.js (hunk 6 above)
Comment (brianjgeiger) [inline website/addons/wiki/static/wikiPage.js]: I've added the visibility checking to the rendering function.

**Comment by JeffSpies** [comment]
There's a good youtube sizer. @icereval, what was that library you used in
the past?

File: website/static/js/pages/wiki-edit-page.js [hunk 9]
```diff
@@ -1,16 +1,153 @@
...
+            // this = all the column elements; an array
+            // title = Text of the button
+            // thisbtn = $(this);
+            // thisbtn.hasClass('btn-primary')
 
-ShareJSDoc('.wiki', url, metadata);
\ No newline at end of file
+            $('body').trigger('togglePanel', [
+                title.toLowerCase(),
+                thisbtn.hasClass('btn-primary')
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: Relying on the the lowercased name of the buttons and the CSS class applied to them makes me nervous--is there a better way for `osfPanel` to pass the "enabled" state of the button to the `onclick` handler? ping @caneruguz

File: website/static/js/pages/wiki-edit-page.js [hunk 10]
```diff
@@ -1,16 +1,153 @@
...
+            ]);
+
+            if (typeof editor !== 'undefined') { ace.edit(editor).resize(); }
+
+        }
+    });
+    $('.openNewWiki').click(function () {
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: Is there a reason to prefer the imperative jQuery interface for modals rather than using `data-target="newWiki"`? I find the latter more concise and robust (one less CSS selector to depend on).

File: website/static/js/pages/wiki-edit-page.js [hunk 11]
```diff
@@ -1,16 +1,153 @@
...
+        $('#newWiki').modal('show');
+    });
+    $('.openDeleteWiki').click(function () {
+        $('#deleteWiki').modal('show');
+    });
+    var panelToggle = $('.panel-toggle'),
+        panelExpand = $('.panel-expand');
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: > Declare one variable per var statement

https://github.com/felixge/node-style-guide#declare-one-variable-per-var-statement

File: website/static/js/pages/wiki-edit-page.js (hunk 9 above)
Comment (caneruguz) [inline website/static/js/pages/wiki-edit-page.js]: Yes there are multiple ways.  The column that is part of this plugin has an attribute called "data-osf-toggle" which is "on" when visible "off" when not visible. This attribtue is on the column that is being toggled. 

However I don't think there is a lot of reason to get nervous. Both these things are set programmatically, (both the text of the button and the classes for on and off) so the user does not have to remember to put these if the plugin is set properly.

File: website/static/js/pages/wiki-edit-page.js [hunk 12]
```diff
@@ -1,16 +1,153 @@
...
+    $('.panel-collapse').on('click', function () {
+        var el = $(this).closest('.panel-toggle');
+        el.children('.wiki-panel.hidden-xs').hide();
+        panelToggle.removeClass('col-sm-3').addClass('col-sm-1');
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: Shouldn't this logic happen in the osfPanel plugin? ping @caneruguz

Comment (brianjgeiger) [inline website/static/js/pages/wiki-edit-page.js]: I suspect this is because of the collapsible menu, yes? That interrupted the cleanliness of the column design.

**Comment by brianjgeiger** [comment]
We should also have something for Vimeo, which we can integrate at any point.

File: website/static/js/pages/wiki-edit-page.js (hunk 12 above)
Comment (caneruguz) [inline website/static/js/pages/wiki-edit-page.js]: Yes, the toggle plugin is not for collapsing, I considered this used case to be specific to OSF and should be handled on our side, to also help with future uses.

File: website/static/js/pages/wiki-edit-page.js (hunk 11 above)
Comment (caneruguz) [inline website/static/js/pages/wiki-edit-page.js]: This is against lint requirements though, is it not?

File: website/static/js/pages/wiki-edit-page.js (hunk 10 above)
Comment (caneruguz) [inline website/static/js/pages/wiki-edit-page.js]: Yes I agree with this, I prefer it that way too but I think there was a complication here related to data-target links being hidden on load where it didn't open the modals; because of the toggly(!) nature of this page I went back to using explicit programmatic calls to the modals I think . It's quite likely though that was because I misunderstood the problem.

File: website/addons/wiki/views.py [hunk 13]
```diff
@@ -245,14 +219,45 @@ def project_wiki_delete(auth, wname, **kwargs):
...
+    wiki_name = (wname or '').strip()
     wiki_key = to_mongo_key(wiki_name)
     wiki_page = node.get_wiki_page(wiki_name)
+    toc = _serialize_wiki_toc(node, auth=auth)
+    can_edit = node.has_permission(auth.user, 'write') and not node.is_registration
+    versions = _get_wiki_versions(node, wiki_name, anonymous=anonymous)
+
+    # Parse URL to determine default panel display
+    view_settings = {}
+    param_list = path.lower().split('/') if path else []
```
Comment (sloria) [inline website/addons/wiki/views.py]: I'm thinking we'll want to switch to using query params (`?mode=view,edit`) for this. Using the path like this not only complicated the view logic, but it also means we will need to be maintain route persistence/backwards-compatibility for the new routes.

File: website/static/js/markdown.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+var hljs = require('highlight.js');
+require('highlight-css');
+var MarkdownIt = require('markdown-it');
+
+highlighter = function (str, lang) {
```
Comment (sloria) [inline website/static/js/markdown.js]: This variable must be defined.

File: website/static/js/markdown.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
+var hljs = require('highlight.js');
```
Comment (sloria) [inline website/static/js/markdown.js]: Missing `'use strict';`.

File: website/static/js/markdown.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+var markdown = new MarkdownIt('commonmark', {
+    highlight: highlighter
+})
+    .use(require('markdown-it-video'))
+    .use(require('markdown-it-toc'))
+    .use(require('markdown-it-sanitizer'));
+
+
+// Fast markdown renderer for active editing to prevent slow loading/rendering tasks
+var markdown_quick = new MarkdownIt(('commonmark'), {
```
Comment (sloria) [inline website/static/js/markdown.js]: Use `lowerCamelCase` for function names.

File: website/static/js/pages/wiki-edit-page.js [hunk 17]
```diff
@@ -1,16 +1,153 @@
+var $ = require('jquery');
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: Missing `'use strict';`

**Comment by rliebz** [comment]
In reference to https://github.com/CenterForOpenScience/osf.io/pull/1858#issuecomment-75583500, I would guess this is related to differences in how newline characters are being handled. When doing dirty checking, I had been running into an issue where /r/n and /n were being generated in different places to describe the same text.

File: website/addons/wiki/views.py [hunk 18]
```diff
@@ -245,14 +219,45 @@ def project_wiki_delete(auth, wname, **kwargs):
...
+                raise WIKI_PAGE_NOT_FOUND_ERROR
+            elif version_number == len(versions):
+                view_settings[param] = True
+            else:
+                view_settings[param] = version_number
+        else:
+            view_settings[param] = True
+
+    # Build styles for default panel display
+    panels = ['view', 'edit', 'compare']
```
Comment (sloria) [inline website/addons/wiki/views.py]: Can this logic (L251-260) happen client-side?

File: website/static/js/pages/wiki-edit-page.js [hunk 19]
```diff
@@ -1,16 +1,153 @@
...
+
+    var options = ctx.viewSettings;
+    var menu = 'menu' in options;
+    var edit = 'edit' in options;
+    var compare = 'compare' in options;
+    var view = 'view' in options;
+    // TODO: toggle menu based on url
+    // TODO: Change version select on toggle
+
+    if (edit) {
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: @brianjgeiger This logic should be handled by the parent ViewModel in `wikiPage.js`.

File: website/addons/wiki/views.py (hunk 18 above)
Comment (rliebz) [inline website/addons/wiki/views.py]: Yes and no. Because this logic is used to define which columns are visible on page load and their widths, I'm not sure there's a way to do this on the client without flashing the columns on page load.

File: website/addons/wiki/views.py (hunk 13 above)
Comment (brianjgeiger) [inline website/addons/wiki/views.py]: Yeah, I tend to agree. We probably want three main views (/view/, /edit/, and /compare/) and then allow changes to those base views with query parameters. The implication is that there would be a hierarchy of base views. For example, /edit/ would make the edit and view columns open, /view/ would be just the view column, and /compare/ would be view and compare. If any of those states are active, the urls would match that. Having all three would be edit with a ?with=compare (and maybe something about the menu state), or if you wanted edit and compare without view you might do ?with=compare&without=view. Alternately, we could have any deviation list out the totality of the column state so we don't have to do plus/minus style routes. 

What I'm saying is we should chat for a couple of minutes.

File: website/templates/projectGridTemplates.html [hunk 20]
```diff
@@ -49,7 +49,7 @@
                             <fieldset class="project-detail-fields">
                                 <div class="row">
                                     <div class="col-xs-7 col-sm-12 col-md-7 col-lg-8">
-                                        <input class="typeahead" id="input{{theItem.node_id}}" type="text" placeholder="Name of project to find" autofocus>
+                                        <input class="typeahead" id="input{{theItem.node_id}}" style="height: 32px;" type="text" placeholder="Name of project to find" autofocus>
```
Comment (sloria) [inline website/templates/projectGridTemplates.html]: Why was this changed?

**Comment by sloria** [comment]
1st pass finished. :point_up:

File: website/templates/projectGridTemplates.html (hunk 20 above)
Comment (brianjgeiger) [inline website/templates/projectGridTemplates.html]: @caneruguz ?

File: website/static/js/pages/wiki-edit-page.js (hunk 19 above)
Comment (brianjgeiger) [inline website/static/js/pages/wiki-edit-page.js]: This is initial state stuff, which will not be how the rest of the ViewModel works. We could put it in the initialization function but probably not the ViewModel per se.

File: website/static/js/markdown.js (hunk 15 above)
Comment (brianjgeiger) [inline website/static/js/markdown.js]: Fixed.

File: website/static/js/pages/wiki-edit-page.js (hunk 17 above)
Comment (brianjgeiger) [inline website/static/js/pages/wiki-edit-page.js]: Fixed.

File: website/templates/projectGridTemplates.html (hunk 20 above)
Comment (caneruguz) [inline website/templates/projectGridTemplates.html]: This seems to be a mistaken commit to here, I was probably working on fixing something else. Current master does not have this change so I would discard this.

File: website/addons/wiki/views.py (hunk 13 above)
Comment (brianjgeiger) [inline website/addons/wiki/views.py]: Okay, @sloria and I talked. We've decided to simplify the URL structure quite a bit and get rid of the implicit states. The main URL that you go to will be the wiki view url `/<pid>/wiki/<wid>/`. We'll redirect from  `/edit/` and `/history/` to the new url structure. 

For everything else, we'll use query parameters/ `?compare&view&edit&menu`. If there's anything at all in query parameters, then view has to be there otherwise it means the column is disabled. Also, compare and view can have optional parameters to indicate version, so `?compare=5&view=preview`, for example.

I think we'll still have implicit versions for compare that are situational, or we need a special `?compare=-1` option (or something more verbose) to indicate that we want the most recent non-current version. Otherwise, by default view is current unless edit is on at the beginning in which case it's preview, and compare by default is most recent (i.e. -1) unless edit is on and view set to preview in which case it's current. (This may need to be simplified before everything is over).

**Comment by icereval** [comment]
Bootstrap's responsive embed works well.

http://getbootstrap.com/components/#responsive-embed

File: website/static/js/pages/wiki-edit-page.js (hunk 10 above)
Comment (caneruguz) [inline website/static/js/pages/wiki-edit-page.js]: I removed this and everything seems to be working. It was probably due to an unrelated issue. Will be part of my PR to Brian's branch.

**Comment by brianjgeiger** [comment]
Responsiveness for videos is fixed.

**Comment by brianjgeiger** [comment]
Snippets added

**Comment by JeffSpies** [comment]
How did you do it? Did you talk to Michael?

File: website/addons/wiki/views.py (hunk 18 above)
Comment (sloria) [inline website/addons/wiki/views.py]: To avoid the flash, we could hide the columns until the javascript is initialized (like we do with the `scripted` class).

Comment (rliebz) [inline website/addons/wiki/views.py]: I believe @JeffSpies wants wiki content visible before javascript is run so that it comes up on searches.

**Comment by brianjgeiger** [comment]
@JeffSpies Michael responded above to say that bootstrap has a responsive video component. I used those classes.

File: website/addons/wiki/views.py (hunk 13 above)
Comment (brianjgeiger) [inline website/addons/wiki/views.py]: And we've decided to add a `previous` special item (like we have with `current` and `preview`) to indicate the last non-current version. This will be used for compare and is necessary to be specific with the query parameters.

**Comment by brianjgeiger** [comment]
I think I fixed the problem with the comparison as well by normalizing the newlines (h/t @rliebz).

File: website/addons/wiki/static/wikiPage.js [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+    self.renderedCompare = ko.observable('');
+    self.allowMathjaxification = ko.observable(true);
+    self.allowFullRender = ko.observable(true);
+
+    self.currentURL = ko.computed(function() {
+        var url = self.pageURL;
+
+        // Default view is special cased
+        if (!self.editVis() && self.viewVis() && self.viewVersion() === 'current' && !self.compareVis() && self.menuVis()) {
+            history.replaceState({}, '', url);  // jshint ignore: line
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: Use `window.history` to clarify where the `history` variable is coming from.

File: website/addons/wiki/static/wikiPage.js [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+            return;
+        }
+
+        var paramPrefix = '?';
+
+        if (self.editVis()) {
+            url += paramPrefix + 'edit';
+            paramPrefix = '&';
```
Comment (sloria) [inline website/addons/wiki/static/wikiPage.js]: Use `$.param` to build query strings.

File: website/addons/wiki/templates/wiki_widget.mako [hunk 23]
```diff
@@ -1,8 +1,20 @@
 <%inherit file="project/addon/widget.mako"/>
 <%page expression_filter="h"/>
 
-% if content:
-    <div>${content | n}</div>
-% else:
-    <p><em>No wiki content</em></p>
-% endif
+<div id="markdown-it-render">
```
Comment (sloria) [inline website/addons/wiki/templates/wiki_widget.mako]: Use `camelCase` for CSS Ids.

File: website/static/js/pages/wiki-edit-page.js [hunk 24]
```diff
@@ -1,16 +1,131 @@
...
+        onSize : 'xs',
+        'onclick' : function (title, thisbtn, event ) {
+            // this = all the column elements; an array
+            // title = Text of the button
+            // thisbtn = $(this);
+            // thisbtn.hasClass('btn-primary')
+
+            bodyElement.trigger('togglePanel', [
+                title.toLowerCase(),
+                thisbtn.hasClass('btn-primary')
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: Not sure where my previous comment went, but this is still brittle. We can defer changing this, but it would be better if could get the toggle state of the clicked button in a more robust way, either by exposing `onEnable/onDisable` callbacks and or a single `onToggle` callback which takes `enabled` as an argument. ping @caneruguz

**Comment by sloria** [comment]
Pass finished. :whale:

**Comment by sloria** [comment]
Suggestions for refactoring `project_wiki_view` and the template were made with @rliebz in person.

File: website/addons/wiki/static/wikiPage.js (hunk 22 above)
Comment (brianjgeiger) [inline website/addons/wiki/static/wikiPage.js]: Talked to @sloria; we are going to defer this because using JQuery param would make messier URLs.

**Comment by brianjgeiger** [comment]
Second pass finished. :bird:

File: website/addons/wiki/routes.py [hunk 25]
```diff
@@ -112,17 +120,14 @@
...
-        Rule([
-            '/project/<pid>/wiki/<wname>/compare/<int:wver>/',
-            '/project/<pid>/node/<nid>/wiki/<wname>/compare/<int:wver>/',
-        ], 'get', views.project_wiki_compare, json_renderer),
-
         # Delete : DELETE
         Rule([
             '/project/<pid>/wiki/<wname>/',
             '/project/<pid>/node/<nid>/wiki/<wname>/',
+            '/project/<pid>/wiki/<wname>/edit/',
```
Comment (sloria) [inline website/addons/wiki/routes.py]: Why allow `DELETE` requests to edit and compare routes? For backwards-compat?

Comment (rliebz) [inline website/addons/wiki/routes.py]: This change was made in regards to a intermediary routing system used in development. I will remove these.

File: website/static/css/site.css [hunk 26]
```diff
@@ -2109,3 +2130,96 @@ span#profileFullname{
...
+/* Changes to default icons to be consistent with the grey look of many new ui items */
+.btn-default-alt {
+    background: #EFEFEF;
+}
+.btn-success-alt {
+    background: #D9E8D9;
+}
+
+
+@media (min-width: 1601px) {
```
Comment (sloria) [inline website/static/css/site.css]: This can probably be reduced. I've got my browser maximized on my 13" Macbook Air, and I'm kinda sad that I don't get to enjoy the full-width editing bliss.

Comment (sloria) [inline website/static/css/site.css]: ping @caneruguz

**Comment by sloria** [comment]
Mathjax doesn't seem to be rendering correctly. The text is rendered 3 times:

cursor_and_osf___wiki_redesign_wiki

**Comment by brianjgeiger** [comment]
Confirmed on the mathjaxing. I wonder what happened there.

File: website/static/css/site.css (hunk 26 above)
Comment (caneruguz) [inline website/static/css/site.css]: Wait why is this in the entire site.css? It means this will apply to everything on the website. 

Anything below this is bootstrap's regular sizing for fixed containers, which causes the constained widths. We can reduce this but this css should be in the wiki.mako or someplace where only wiki is loading.

Comment (brianjgeiger) [inline website/static/css/site.css]: Take a look at the next line.

Comment (caneruguz) [inline website/static/css/site.css]: Ah I see, there is a specific class there. But this does not expand the container in the navigation and project menu, they will be the regular size while the wiki will continue to expand. As long as this is the desired outcome.

Comment (brianjgeiger) [inline website/static/css/site.css]: It was. We went back and forth, and may change out minds, but that seemed the better path for now.

**Comment by sloria** [comment]
Merged into `release/0.27.0`
